### PR TITLE
Fix Promise.reject

### LINKED ISSUE #23416: Promise.reject should be callable with no parameter
**TypeScript Version:**  master

**Search Terms:** 

**Code**

```ts
    reject(reason: any): Promise<never>;
    reject<T>(reason: any): Promise<T>;
    resolve<T>(value: T | PromiseLike<T>): Promise<T>;
    resolve(): Promise<void>;
```

**Expected behavior:**

```ts
    reject<T = never>(reason?: any): Promise<T>;
```

**Actual behavior:**

```ts
    reject(reason: any): Promise<never>;
    reject<T>(reason: any): Promise<T>;
```

**Playground Link:**  

**Related Issues:**

### PR DESCRIPTION
**Pull request by falsandtru**
Fixes #23416

@rbuckton

### CODE REVIEW

File: src/lib/es2015.promise.d.ts [hunk 1]
```diff
@@ -177,14 +177,7 @@ interface PromiseConstructor {
...
      */
-    reject(reason: any): Promise<never>;
-
-    /**
-     * Creates a new rejected promise for the provided reason.
-     * @param reason The reason the promise was rejected.
-     * @returns A new rejected Promise.
-     */
-    reject<T>(reason: any): Promise<T>;
+    reject<T = never>(reason?: any): Promise<T>;
```
Comment (rbuckton) [inline src/lib/es2015.promise.d.ts]: I would not recommend making `reason` optional. Yes, `undefined` is permitted but it is generally a bad idea to throw `undefined` (which is effectively what this is doing). Requiring the parameter helps catch a certain class of errors, namely an "I forgot to fill in a reason" typo. However, if you really and truly want to create a promise rejected with `undefined`, you can still be explicit about it (i.e. `Promise.reject(undefined)`).

Other than that, I'm perfectly fine with collapsing the two declarations into a single declaration with a default type argument.

Comment (falsandtru) [reply src/lib/es2015.promise.d.ts]: But I feel requiring the reason parameter is not the purpose of #23416. And it doesn't reflect the ES spec. @mhegazy Do you want to make a reason required?

Comment (rbuckton) [reply src/lib/es2015.promise.d.ts]: After a quick review it seems that we allow the reason to be elided in the `reject` callback passed as part of the `Promise` constructor.

**Comment by falsandtru** [comment]
Thanks for your approval.

**Pull request merged**